# Neighborhoods in Toronto

### **Part 1: Scrape Wikipedia page**

In [1]:
import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup

Waiting for a Spark session to start...
Spark Initialization Done! ApplicationId = app-20190601122512-0005
KERNEL_ID = d948bdfa-0730-4c0d-b628-808cbdc40455


##### *Assign the link of the website to variable named website_url*

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

##### *Load website_url, turn into soup and get the tables*

In [3]:
soup = BeautifulSoup(website_url,'lxml')
tables = soup.find('table',{'class':'wikitable sortable'})

##### *Extract the header*

In [4]:
header = [head.findAll(text=True)[0].strip() for head in tables.find_all("th")]
header

['Postcode', 'Borough', 'Neighbourhood']

##### *Extract the columns we want*

In [5]:
data   = [[td.findAll(text=True)[0].strip() for td in tr.find_all("td")]
              for tr in tables.find_all("tr")]
data    = [row for row in data if len(row) == 3]

##### *Convert the header and data into Pandas DataFrame*

In [6]:
df = pd.DataFrame(data, columns=header)
df.rename(columns = {"Neighbourhood": "Neighborhood"}, inplace = True) 
df.head(20)

Postcode           Borough      Neighborhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
9       M8A      Not assigned      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern
13      M2B      Not assigned      Not assigned
14      M3B        North York   Don Mills North
15      M4B         East York  Woodbine Gardens
16      M4B         East York     Parkview Hill
17      M5B  Downtown Toronto           Ryerson
18      M5B  Downtown Toronto   Garden District
19      M6B        North York         Glencairn

##### *Remove cells with a borough that is 'Not assigned'*

In [7]:
indexNames = df[ df['Borough'] == 'Not assigned' ].index
df.drop(indexNames , inplace=True)

##### *If a cell has a borough but a 'Not assigned' neighborhood, then the neighborhood will be the same as the borough*

In [8]:
df['Neighborhood'] = np.where(df['Neighborhood'] == 'Not assigned', df['Borough'], df['Neighborhood'])

##### *Combine the rows with the same postcode and borough into one row and with the neighborhoods separated with a comma*

In [9]:
df = df.groupby(['Postcode','Borough'], sort=False)['Neighborhood'].apply(lambda x: ', '.join(x)).reset_index()
df.head(20)

Postcode           Borough  \
0       M3A        North York   
1       M4A        North York   
2       M5A  Downtown Toronto   
3       M6A        North York   
4       M7A      Queen's Park   
5       M9A         Etobicoke   
6       M1B       Scarborough   
7       M3B        North York   
8       M4B         East York   
9       M5B  Downtown Toronto   
10      M6B        North York   
11      M9B         Etobicoke   
12      M1C       Scarborough   
13      M3C        North York   
14      M4C         East York   
15      M5C  Downtown Toronto   
16      M6C              York   
17      M9C         Etobicoke   
18      M1E       Scarborough   
19      M4E      East Toronto   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Harbourfront, Regent Park  
3                    Lawrence Heights, Lawrence Manor  
4                                        Queen's Park  
5                                    Islington Avenue  
6                                      Rouge, Malvern  
7                                     Don Mills North  
8                     Woodbine Gardens, Parkview Hill  
9                            Ryerson, Garden District  
10                                          Glencairn  
11  Cloverdale, Islington, Martin Grove, Princess ...  
12             Highland Creek, Rouge Hill, Port Union  
13                   Flemingdon Park, Don Mills South  
14                                   Woodbine Heights  
15                                     St. James Town  
16                                 Humewood-Cedarvale  
17  Bloordale Gardens, Eringate, Markland Wood, Ol...  
18                  Guildwood, Morningside, West Hill  
19                                        The Beaches

##### *Print the number of rows of the dataframe*

In [10]:
df.shape

(103, 3)

### **Part 2: Get the latitude and the longitude coordinates of each neighborhood.**

##### *convert Geospatial_data csv to dataframe*

In [11]:
data = pd.read_csv("http://cocl.us/Geospatial_data") 

##### *rename column Postcode to Postal Code*

In [12]:
df.rename(columns = {'Postcode':'Postal Code'}, inplace=True)
df_new = pd.concat([df, data], axis=1)

##### *merge Toronto neighbourhood dataframe from Part 1 and Geospatial_data dataframe together*

In [13]:
df_new = pd.merge(df, data, on = 'Postal Code')
df_new.head(10)

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3         M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4         M7A      Queen's Park                      Queen's Park  43.662301   
5         M9A         Etobicoke                  Islington Avenue  43.667856   
6         M1B       Scarborough                    Rouge, Malvern  43.806686   
7         M3B        North York                   Don Mills North  43.745906   
8         M4B         East York   Woodbine Gardens, Parkview Hill  43.706397   
9         M5B  Downtown Toronto          Ryerson, Garden District  43.657162   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494  
5 -79.532242  
6 -79.194353  
7 -79.352188  
8 -79.309937  
9 -79.378937

### **Part 3: Explore and cluster the neighborhoods in Toronto.**

In [14]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

#!conda install -c conda-forge folium --yes
!pip install folium
import folium # plotting library

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

Solving environment: done

## Package Plan ##

  environment location: /opt/ibm/conda/miniconda36

  added / updated specs:
    - geopy


The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.49-py_0
  geopy              conda-forge/noarch::geopy-1.20.0-py_0

The following packages will be UPDATED:

  ca-certificates    pkgs/main::ca-certificates-2019.1.23-0 --> conda-forge::ca-certificates-2019.3.9-hecc5488_0

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi                                         pkgs/main --> conda-forge
  conda                                           pkgs/main --> conda-forge
  openssl              pkgs/main::openssl-1.1.1b-h7b6447c_1 --> conda-forge::openssl-1.1.1b-h14c3975_1


Preparing transaction: done
Verifying transaction: failed

EnvironmentNotWritableError: The current user does not have write permissions to the target environment.
  environment location: /opt/ibm/conda/mi

##### *Get the coordinate of Toronto using geolocator*

In [15]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


##### Create a map of Toronto

In [16]:
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)

#add neighborhood markers to the Toronto map
for lat, long, bor, neigh in zip(df_new['Latitude'], df_new['Longitude'], 
                                 df_new['Borough'], df_new['Neighborhood']):
    label = '{}, {}'.format(neigh, bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius = 7, 
        popup = label,
        color = 'red',
        fill = True,
        fill_color = 'white',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)
        
map_toronto

In our analysis, we will choose "Downtown Toronto" as the borough of interest and then cluster the neighborhoods around it including those in the boroughs with 'Toronto' in their names, e.g. East Toronto, Central Toronto.

Therefore, int the code below, we will sort the Boroughs that contain the word 'Toronto'.

In [17]:
new_data = df_new[df_new['Borough'].str.contains("Toronto")].reset_index(drop=True)
new_data.head(10)

Postal Code           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M4E      East Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M6H      West Toronto   
9         M5J  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  
0                          Harbourfront, Regent Park  43.654260 -79.360636  
1                           Ryerson, Garden District  43.657162 -79.378937  
2                                     St. James Town  43.651494 -79.375418  
3                                        The Beaches  43.676357 -79.293031  
4                                        Berczy Park  43.644771 -79.373306  
5                                 Central Bay Street  43.657952 -79.387383  
6                                           Christie  43.669542 -79.422564  
7                           Adelaide, King, Richmond  43.650571 -79.384568  
8                       Dovercourt Village, Dufferin  43.669005 -79.442259  
9  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752

##### Get Co-ordinates of Downtown Toronto

In [18]:
address = 'Downtown Toronto, TO'

geolocator = Nominatim()
location = geolocator.geocode(address)
neighborhood_latitude = location.latitude
neighborhood_longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(neighborhood_latitude, neighborhood_longitude))

/opt/ibm/conda/miniconda36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Downtown Toronto are 43.65145205, -79.3620078930752.


##### Lets create a map of Downtown Toronto and its surrounding Neighborhoods

In [19]:
map_dwtor = folium.Map(location=[neighborhood_latitude, neighborhood_longitude], zoom_start=11)

# Showing Downt. Toronto as Red on the map ....
folium.CircleMarker([neighborhood_latitude, neighborhood_longitude], radius=10, popup=label, color='Red', fill=True, fill_color='Red', fill_opacity=0.9,).add_to(map_dwtor)

# ... and the Neighbourhoods as Blue.
for lat, lng, label in zip(new_data['Latitude'], new_data['Longitude'], new_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_dwtor)  
    
map_dwtor

##### Define Foursquare Credentials and Version

In [20]:
# The code was removed by Watson Studio for sharing.

##### Let's get the top 100 venues that are in Downtown Toronto within a radius of 500 meters
First, let's create the GET request URL. Name your URL **url**.

In [21]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

##### Send the GET request

In [22]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cf26ee36a60714939f582b9'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Downtown Toronto',
  'headerFullLocation': 'Downtown Toronto, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.655952054500005,
    'lng': -79.35580018346758},
   'sw': {'lat': 43.6469520455, 'lng': -79.36821560268282}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4c3e1eaa6faac9b66dc60d76',
       'name': 'Distillery Sunday Market',
       'location': {'address': '1 Trinity St',
        'lat': 43.650074989330655,
        'lng': -79.36183171318665,
        'labeledLatLngs': [{'label': 'display',
    

##### Create function that extracts the category of the venue

In [23]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

##### Clean the json and structure it into a pandas dataframe.

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(10)

name      categories        lat        lng
0          Distillery Sunday Market  Farmers Market  43.650075 -79.361832
1                              Arvo     Coffee Shop  43.649963 -79.361442
2                          Cacao 70    Dessert Shop  43.650067 -79.360723
3                         Starbucks     Coffee Shop  43.651327 -79.364329
4                  Roselle Desserts          Bakery  43.653447 -79.362017
5  The Distillery Historic District   Historic Site  43.650244 -79.359323
6                     Tandem Coffee     Coffee Shop  43.653559 -79.361809
7            Parliament Square Park            Park  43.650264 -79.362195
8                    Rooster Coffee     Coffee Shop  43.651900 -79.365609
9                              LCBO    Liquor Store  43.650982 -79.365361

In [25]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


##### Create a function to repeat the same process to all the neighborhoods' name contains 'Toronto'

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

##### create a new dataframe called toronto_venues

In [27]:
toronto_venues = getNearbyVenues(names=new_data['Neighborhood'],
                                   latitudes=new_data['Latitude'],
                                   longitudes=new_data['Longitude']
                                  )


Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

##### Check the size of the resulting dataframe

In [28]:
print(toronto_venues.shape)
toronto_venues.head(10)

(1690, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   
5  Harbourfront, Regent Park               43.65426              -79.360636   
6  Harbourfront, Regent Park               43.65426              -79.360636   
7  Harbourfront, Regent Park               43.65426              -79.360636   
8  Harbourfront, Regent Park               43.65426              -79.360636   
9  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                              Body Blitz Spa East       43.654735   
4                               Morning Glory Cafe       43.653947   
5                                   Impact Kitchen       43.656369   
6                           Figs Breakfast & Lunch       43.655675   
7                                  Corktown Common       43.655618   
8                         Dominion Pub and Kitchen       43.656919   
9                 The Distillery Historic District       43.650244   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.359874                   Spa  
4       -79.361149        Breakfast Spot  
5       -79.356980            Restaurant  
6       -79.364503        Breakfast Spot  
7       -79.356211                  Park  
8       -79.358967                   Pub  
9       -79.359323         Historic Site

So we can see that there are total 1690 venues in the neighbourhoods in our dataset. The dataframe generated gives interesting information about Venue names like Roselle Desserts or Morning Glory Cafe and their corresponding categories like Bakery or Breakfast Spot.

We can use this interesting information to categorize these neigborhoods using k_Means clustering to get insights about business or real estate investments.

Lets see number of venues returned per neighbourhood

##### Check how many venues were returned for each neighborhood

In [29]:
toronto_venues.groupby('Neighborhood').count()
toronto_venues

Neighborhood  \
0                             Harbourfront, Regent Park   
1                             Harbourfront, Regent Park   
2                             Harbourfront, Regent Park   
3                             Harbourfront, Regent Park   
4                             Harbourfront, Regent Park   
5                             Harbourfront, Regent Park   
6                             Harbourfront, Regent Park   
7                             Harbourfront, Regent Park   
8                             Harbourfront, Regent Park   
9                             Harbourfront, Regent Park   
10                            Harbourfront, Regent Park   
11                            Harbourfront, Regent Park   
12                            Harbourfront, Regent Park   
13                            Harbourfront, Regent Park   
14                            Harbourfront, Regent Park   
15                            Harbourfront, Regent Park   
16                            Harbourfront, Regent Park   
17                            Harbourfront, Regent Park   
18                            Harbourfront, Regent Park   
19                            Harbourfront, Regent Park   
20                            Harbourfront, Regent Park   
21                            Harbourfront, Regent Park   
22                            Harbourfront, Regent Park   
23                            Harbourfront, Regent Park   
24                            Harbourfront, Regent Park   
25                            Harbourfront, Regent Park   
26                            Harbourfront, Regent Park   
27                            Harbourfront, Regent Park   
28                            Harbourfront, Regent Park   
29                            Harbourfront, Regent Park   
...                                                 ...   
1660                               Church and Wellesley   
1661                               Church and Wellesley   
1662                               Church and Wellesley   
1663                               Church and Wellesley   
1664                               Church and Wellesley   
1665                               Church and Wellesley   
1666                               Church and Wellesley   
1667                               Church and Wellesley   
1668                               Church and Wellesley   
1669                               Church and Wellesley   
1670                               Church and Wellesley   
1671                               Church and Wellesley   
1672  Business Reply Mail Processing Centre 969 Eastern   
1673  Business Reply Mail Processing Centre 969 Eastern   
1674  Business Reply Mail Processing Centre 969 Eastern   
1675  Business Reply Mail Processing Centre 969 Eastern   
1676  Business Reply Mail Processing Centre 969 Eastern   
1677  Business Reply Mail Processing Centre 969 Eastern   
1678  Business Reply Mail Processing Centre 969 Eastern   
1679  Business Reply Mail Processing Centre 969 Eastern   
1680  Business Reply Mail Processing Centre 969 Eastern   
1681  Business Reply Mail Processing Centre 969 Eastern   
1682  Business Reply Mail Processing Centre 969 Eastern   
1683  Business Reply Mail Processing Centre 969 Eastern   
1684  Business Reply Mail Processing Centre 969 Eastern   
1685  Business Reply Mail Processing Centre 969 Eastern   
1686  Business Reply Mail Processing Centre 969 Eastern   
1687  Business Reply Mail Processing Centre 969 Eastern   
1688  Business Reply Mail Processing Centre 969 Eastern   
1689  Business Reply Mail Processing Centre 969 Eastern   

      Neighborhood Latitude  Neighborhood Longitude  \
0                 43.654260              -79.360636   
1                 43.654260              -79.360636   
2                 43.654260              -79.360636   
3                 43.654260              -79.360636   
4                 43.654260              -79.360636   
5                 43.654260              -79.360636   

In [30]:
category = toronto_venues.groupby('Venue Category').count()
category.sort_values('Neighborhood', ascending  = False)

Neighborhood  Neighborhood Latitude  \
Venue Category                                                         
Coffee Shop                               143                    143   
Café                                       94                     94   
Restaurant                                 49                     49   
Italian Restaurant                         47                     47   
Bakery                                     45                     45   
Hotel                                      39                     39   
Bar                                        37                     37   
Pizza Place                                34                     34   
Park                                       31                     31   
Japanese Restaurant                        27                     27   
American Restaurant                        24                     24   
Gastropub                                  24                     24   
Sandwich Place                             23                     23   
Seafood Restaurant                         23                     23   
Breakfast Spot                             21                     21   
Gym                                        21                     21   
Steakhouse                                 21                     21   
Sushi Restaurant                           20                     20   
Pub                                        19                     19   
Burger Joint                               19                     19   
Clothing Store                             19                     19   
Vegetarian / Vegan Restaurant              19                     19   
Beer Bar                                   17                     17   
Thai Restaurant                            17                     17   
Bookstore                                  16                     16   
Greek Restaurant                           16                     16   
Fast Food Restaurant                       16                     16   
Deli / Bodega                              15                     15   
Diner                                      15                     15   
Ice Cream Shop                             15                     15   
...                                       ...                    ...   
Massage Studio                              1                      1   
Garden Center                               1                      1   
Persian Restaurant                          1                      1   
Organic Grocery                             1                      1   
Pool                                        1                      1   
Optical Shop                                1                      1   
Fruit & Vegetable Store                     1                      1   
Molecular Gastronomy Restaurant             1                      1   
Filipino Restaurant                         1                      1   
Food                                        1                      1   
Flea Market                                 1                      1   
Malay Restaurant                            1                      1   
Hotpot Restaurant                           1                      1   
Skating Rink                                1                      1   
Climbing Gym                                1                      1   
Flower Shop                                 1                      1   
German Restaurant                           1                      1   
Cajun / Creole Restaurant                   1                      1   
Camera Store                                1                      1   
Chocolate Shop                              1                      1   
Lingerie Store                              1                      1   
College Arts Building                       1                      1   
Post Office                                 1                      1   
S

We can see from above that Coffe Shop, Cafe, Restaurant, Italian Restaurant, Bakery, Hotel are the most frequently observed venues. The occurence of any of these venues in any neighbourhood will define the category of that neighborhood. Hence we will cluster the neighborhoods on the basis of the frequency of the venue categories.

##### Let's find out how many unique categories can be curated from all the returned venue

In [31]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 234 uniques categories.


### Analyze Each Neighborhood

In [32]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(10)

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   
5            0               0                  0        0   
6            0               0                  0        0   
7            0               0                  0        0   
8            0               0                  0        0   
9            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   
5                   0             0               0                0   
6                   0             0               0                0   
7                   0             0               0                0   
8                   0             0               0                0   
9                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Thrift / Vintage Store  \
0                 0                    0  ...                       0   
1                 0                    0  ...                       0   
2                 0                    0  ...                       0   
3                 0                    0  ...                       0   
4                 0                    0  ...                       0   
5                 0                    0  ...                       0   
6                 0                    0  ...                       0   
7                 0                    0  ...                       0   
8                 0                    0  ...                       0   
9                 0                    0  ...                       0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   
5                 0      0              0                              0   
6                 0      0              0                              0   
7                 0      0              0                              0   
8                 0      0              0                              0   
9                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wings Joint  \
0                 0                      0         0            0   
1                 0                      0         0            0   
2                 0                      0         0            0   
3                 0                      0         0            0   
4                 0                      0         0            0   
5                 0                      0         0            0   
6                 0                      0         0            0   
7                 0                      0         0            0   
8                 0                      0         0            0   
9                 0                      0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  
5           

In [33]:
toronto_onehot.shape

(1690, 234)

##### Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [34]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.055556   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011364   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011364   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.028571   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.022222   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.016129   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.050000   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.027027   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011364           0.011364  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

In [35]:
toronto_grouped.shape

(38, 234)

We have 38 unique neighbourhoods

##### Let's print each neighborhood along with the top 5 most common venues

In [36]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3                  Bar  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Steakhouse  0.04
3  Italian Restaurant  0.04
4            Beer Bar  0.04


----Brockton, Exhibition Place, Parkdale Village----
            venue  freq
0  Breakfast Spot  0.11
1     Coffee Shop  0.11
2            Café  0.11
3    Climbing Gym  0.05
4             Gym  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0   Airport Service  0.21
1    Ai

##### Write a function to sort the venues in descending order

In [37]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

##### Create the new dataframe and display the top 10 venues for each neighborhood

In [38]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant                   Bar   
1          Cocktail Bar                  Café            Steakhouse   
2                  Café           Coffee Shop                   Gym   
3           Yoga Studio         Auto Workshop            Comic Shop   
4        Airport Lounge      Airport Terminal       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse   American Restaurant          Burger Joint   
1    Italian Restaurant              Beer Bar    Seafood Restaurant   
2          Climbing Gym         Burrito Place               Stadium   
3           Pizza Place            Restaurant               Butcher   
4         Boat or Ferry               Airport    Airport Food Court   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel                Bakery         Cosmetics Shop  
1            Restaurant        Farmers Market                 Bakery  
2                   Bar            Restaurant   Caribbean Restaurant  
3         Burrito Place               Brewery             Skate Park  
4          Airport Gate              Boutique       Sculpture Garden

### Cluster Neighborhoods

##### Run k-means to cluster the neighborhood into 5 clusters

In [39]:
kclusters = 5

grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

##### Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [40]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = new_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(10) # check the last columns!

Postal Code           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M4E      East Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M6H      West Toronto   
9         M5J  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  \
0                          Harbourfront, Regent Park  43.654260 -79.360636   
1                           Ryerson, Garden District  43.657162 -79.378937   
2                                     St. James Town  43.651494 -79.375418   
3                                        The Beaches  43.676357 -79.293031   
4                                        Berczy Park  43.644771 -79.373306   
5                                 Central Bay Street  43.657952 -79.387383   
6                                           Christie  43.669542 -79.422564   
7                           Adelaide, King, Richmond  43.650571 -79.384568   
8                       Dovercourt Village, Dufferin  43.669005 -79.442259   
9  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                Bakery   
1               0           Coffee Shop        Clothing Store   
2               0           Coffee Shop                  Café   
3               0     Health Food Store         Grocery Store   
4               0           Coffee Shop          Cocktail Bar   
5               0           Coffee Shop                  Café   
6               0         Grocery Store                  Café   
7               0           Coffee Shop                  Café   
8               0              Pharmacy           Supermarket   
9               0           Coffee Shop              Aquarium   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                  Park    Mexican Restaurant                    Theater   
1                  Café        Cosmetics Shop  Middle Eastern Restaurant   
2            Restaurant                 Hotel             Cosmetics Shop   
3              Boutique                   Pub                      Diner   
4                  Café            Steakhouse         Italian Restaurant   
5    Italian Restaurant        Sandwich Place               Burger Joint   
6                  Park    Italian Restaurant                 Baby Store   
7       Thai Restaurant                   Bar                 Steakhouse   
8                Bakery           Music Venue                       Pool   
9                 Hotel    Italian Restaurant                       Café   

       6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0                        Pub        Breakfast Spot                       Café   
1           Ramen Restaurant                Bakery                Pizza Place   
2             Breakfast Spot        Clothing Store                     Bakery   
3             Farmers Market    Falafel Restaurant                Event Space   
4                   Beer Bar    Seafood Restaurant                 Restaurant   
5  Middle Eastern Restaurant           Salad Place            Bubble Tea Shop   
6         Athletics & Sports            Restaurant                  Nightclub   
7        American Restaurant          Burger Joint                      Hotel   
8      Portuguese Restaurant        Discount Store  Middle Eastern Restaurant   
9                     Bakery   Fried Chicken Joint                Pizza Place   

  9th Most Common Venue 10th Most Common Venue  
0  Gym / Fitness Center          Historic Site  
1   Sporting Goods Shop        Bubble Tea Shop  
2             Gastropub                  Diner  
3  Ethiopian Restaurant      Electronics Store  
4        Farmers Market                 Bakery  
5                 

##### Visualize the resulting clusters

In [41]:
# create map
map_clusters = folium.Map(location=[neighborhood_latitude, neighborhood_longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine the Cluster

##### Cluster 1

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3       East Toronto               0     Health Food Store   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0         Grocery Store   
7   Downtown Toronto               0           Coffee Shop   
8       West Toronto               0              Pharmacy   
9   Downtown Toronto               0           Coffee Shop   
10      West Toronto               0                   Bar   
11      East Toronto               0      Greek Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13      West Toronto               0        Breakfast Spot   
14      East Toronto               0        Sandwich Place   
15  Downtown Toronto               0           Coffee Shop   
16      East Toronto               0                  Café   
19   Central Toronto               0                 Hotel   
21      West Toronto               0    Mexican Restaurant   
22   Central Toronto               0        Clothing Store   
23   Central Toronto               0           Coffee Shop   
24      West Toronto               0        Breakfast Spot   
25   Central Toronto               0          Dessert Shop   
26  Downtown Toronto               0                  Café   
27      West Toronto               0           Pizza Place   
29  Downtown Toronto               0                  Café   
30   Central Toronto               0           Coffee Shop   
31  Downtown Toronto               0       Airport Service   
33  Downtown Toronto               0           Coffee Shop   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0                  Café   
36  Downtown Toronto               0   Japanese Restaurant   
37      East Toronto               0    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                          Bakery                  Park    Mexican Restaurant   
1                  Clothing Store                  Café        Cosmetics Shop   
2                            Café            Restaurant                 Hotel   
3                   Grocery Store              Boutique                   Pub   
4                    Cocktail Bar                  Café            Steakhouse   
5                            Café    Italian Restaurant        Sandwich Place   
6                            Café                  Park    Italian Restaurant   
7                            Café       Thai Restaurant                   Bar   
8                     Supermarket                Bakery           Music Venue   
9                        Aquarium                 Hotel    Italian Restaurant   
10               Asian Restaurant           Men's Store           Coffee Shop   
11                    Coffee Shop        Ice Cream Shop    Italian Restaurant   
12                          Hotel                  Café            Restaurant   
13                           Café           Coffee Shop                   Gym   
14                      Pet Store        Ice Cream Shop           Pizza Place   
15                           Café                 Hotel            Restaurant   
16                    Coffee Shop             Gastropub   American Restaurant   
19                           Park     Food & Drink Shop        Breakfast Spot   
21                           Café                   Bar         Grocery Store   
22                    Coffee Shop            Bagel Shop             Gift Shop   
23                           Café        Sandwich Place           Pizza Place   
24                      Gift Shop    Italian Restaurant          Dessert Shop   
25                    Pizza Place        

##### Cluster 2

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18         Women's Store        Farmers Market    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

##### Cluster 3

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Central Toronto               3    Mexican Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20         Women's Store                 Diner    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20           Event Space  Ethiopian Restaurant      Electronics Store

##### Cluster 4

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17   Central Toronto               4              Bus Line   
32  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Park           Swim School         Women's Store   
32            Playground                 Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17        Farmers Market    Falafel Restaurant           Event Space   
32          Dessert Shop    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
17  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant  
32  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

##### Cluster 5

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []